File: zig/src/revera_test.zig
const std = @import("std");
const revera = @import("revera.zig");

const testing = std.testing;
const gpa = testing.allocator;

test "find and captures" {
    var re = try revera.Regex.compile(gpa, "([a-z]+)([0-9]*)", .{});
    defer re.deinit();

    try testing.expectEqual(3, re.groupCount());
    try testing.expect(try re.isMatch("__abc12__"));

    const m = (try re.find("__abc12__")).?;
    try testing.expectEqual(2, m.start);
    try testing.expectEqual(7, m.end);
    try testing.expectEqualStrings("abc12", m.text());
    try testing.expectEqual(5, m.len());
    try testing.expect(!m.isEmpty());

    var caps = (try re.captures("__abc12__")).?;
    defer caps.deinit();
    try testing.expectEqual(3, caps.len());
    try testing.expectEqualStrings("abc", caps.get(1).?.text());
    try testing.expectEqualStrings("12", caps.get(2).?.text());
}

test "a group that took no part reads as null" {
    var re = try revera.Regex.compile(gpa, "(a)|(b)", .{});
    defer re.deinit();

    var caps = (try re.captures("a")).?;
    defer caps.deinit();
    try testing.expect(caps.get(1) != null);
    try testing.expect(caps.get(2) == null);
    try testing.expect(caps.get(9) == null);
}

test "no match reads as null" {
    var re = try revera.Regex.compile(gpa, "z+", .{});
    defer re.deinit();

    try testing.expect(!try re.isMatch("abc"));
    try testing.expect((try re.find("abc")) == null);
    try testing.expect((try re.captures("abc")) == null);

    var it = try re.matches("abc");
    try testing.expect((try it.next()) == null);
}

test "iterators walk every match" {
    var re = try revera.Regex.compile(gpa, "(a+)(b*)", .{});
    defer re.deinit();

    const subject = "aab a aabbb";
    var found: std.ArrayList([]const u8) = .empty;
    defer found.deinit(gpa);
    var it = try re.matches(subject);
    while (try it.next()) |m| {
        try found.append(gpa, m.text());
    }
    try testing.expectEqual(3, found.items.len);
    try testing.expectEqualStrings("aab", found.items[0]);
    try testing.expectEqualStrings("a", found.items[1]);
    try testing.expectEqualStrings("aabbb", found.items[2]);

    var groups = try re.captureMatches("aab a");
    var count: usize = 0;
    while (try groups.next()) |c| {
        var caps = c;
        defer caps.deinit();
        try testing.expectEqual(3, caps.len());
        count += 1;
    }
    try testing.expectEqual(2, count);
}

test "replacement" {
    var re = try revera.Regex.compile(gpa, "(a+)(b*)", .{});
    defer re.deinit();

    const all = try re.replaceAll("xaabyy", "[&:\\2]");
    defer gpa.free(all);
    try testing.expectEqualStrings("x[aab:b]yy", all);

    const one = try re.replaceFirstN("aa bb aa", "X", 1);
    defer gpa.free(one);
    try testing.expectEqualStrings("X bb aa", one);
}

test "options" {
    var icase = try revera.Regex.compile(gpa, "ab+", .{ .case_insensitive = true });
    defer icase.deinit();
    try testing.expect(try icase.isMatch("ABBB"));

    var lines = try revera.Regex.compile(gpa, "^b", .{ .newline_sensitive = true });
    defer lines.deinit();
    try testing.expectEqual(2, (try lines.find("a\nbc")).?.start);

    var shortest = try revera.Regex.compile(gpa, "a+", .{ .shortest_match = true });
    defer shortest.deinit();
    try testing.expectEqual(1, (try shortest.find("aaa")).?.end);

    var plain = try revera.Regex.compile(gpa, "a+", .{ .no_captures = true });
    defer plain.deinit();
    try testing.expect(try plain.isMatch("baa"));
    try testing.expectError(error.NoCaptures, plain.find("baa"));
}

test "locales" {
    const cs = (try revera.Locale.open(gpa, "cs", "")).?;
    var re = try revera.Regex.compile(gpa, "[[.ch.]]", .{ .locale = cs });
    defer re.deinit();
    try testing.expect(try re.isMatch("ch"));
    try testing.expect(try revera.Locale.open(gpa, "xx-not-there", "") == null);

    const names = try revera.Locale.names(gpa);
    defer gpa.free(names);
    try testing.expect(names.len > 1000);
}

test "compile failure reports a position" {
    var at: usize = 0;
    const err = revera.Regex.compile(gpa, "a(", .{ .error_position = &at });
    try testing.expectError(error.InvalidPattern, err);
    try testing.expectEqual(2, at);

    try testing.expectError(
        error.InvalidCharacterClass,
        revera.Regex.compile(gpa, "[[:bogus:]]", .{}),
    );
}

test "contract grows with the input bound" {
    var re = try revera.Regex.compile(gpa, "(a|ab)(c|bcd)(d*)", .{});
    defer re.deinit();

    const big = re.contract(1 << 12);
    try testing.expectEqual(1 << 12, big.max_input);
    try testing.expect(big.heap_bytes > 0);
    try testing.expect(big.stack_bytes > 0);
    try testing.expect(big.one_pass == null);
    try testing.expect(big.solver != null);
    try testing.expect(big.matcher.steps > 0);
    try testing.expect(re.contract(16).steps < big.steps);

    // An absurd bound clamps to the subject limit of the engine.
    try testing.expectEqual((1 << 31) - 1, re.contract(std.math.maxInt(usize)).max_input);

    var simple = try revera.Regex.compile(gpa, "(abc+)", .{});
    defer simple.deinit();
    const one_pass = simple.contract(1000);
    try testing.expect(one_pass.one_pass != null);
    try testing.expect(one_pass.solver == null);
    try testing.expectEqual(37_757, one_pass.heap_bytes);
    try testing.expectEqual(6_144, one_pass.stack_bytes);
    try testing.expectEqual(937_980, one_pass.steps);
}

test "one expression serves several threads" {
    var re = try revera.Regex.compile(gpa, "[0-9]+", .{});
    defer re.deinit();

    const search = struct {
        fn run(r: *const revera.Regex) !void {
            for (0..200) |_| {
                const m = (try r.find("ab 1234 cd")).?;
                try testing.expectEqualStrings("1234", m.text());
            }
        }
    }.run;

    var threads: [4]std.Thread = undefined;
    for (&threads) |*t| {
        t.* = try std.Thread.spawn(.{}, search, .{&re});
    }
    for (threads) |t| {
        t.join();
    }
}

File: ts/test/revera.test.ts
import { test } from "node:test";
import assert from "node:assert/strict";

import { Captures, DUP_MAX, Locale, Regex, RegexError, embeddedLocaleData } from "../src/revera.ts";

test("find and captures", () => {
    const re = new Regex("([a-z]+)([0-9]*)");
    assert.equal(re.capturesLength, 3);
    assert.equal(re.test("__abc12__"), true);

    const m = re.find("__abc12__");
    assert.ok(m);
    assert.equal(m.start, 2);
    assert.equal(m.end, 7);
    assert.equal(m.text, "abc12");
    assert.equal(m.length, 5);
    assert.equal(m.isEmpty, false);

    const caps = re.captures("__abc12__");
    assert.ok(caps);
    assert.equal(caps.length, 3);
    assert.equal(caps.get(0)?.text, "abc12");
    assert.equal(caps.get(1)?.text, "abc");
    assert.equal(caps.get(2)?.text, "12");
    assert.deepEqual([...caps].map((c) => c?.text), ["abc12", "abc", "12"]);
});

test("a group that took no part reads as null", () => {
    const re = new Regex("(a)|(b)");
    const caps = re.captures("a");
    assert.ok(caps);
    assert.ok(caps.get(1));
    assert.equal(caps.get(2), null);
    assert.equal(caps.get(9), null);
});

test("no match reads as null", () => {
    const re = new Regex("z+");
    assert.equal(re.test("abc"), false);
    assert.equal(re.find("abc"), null);
    assert.equal(re.captures("abc"), null);
    assert.deepEqual([...re.matches("abc")], []);
});

test("iterators walk every match", () => {
    const re = new Regex("(a+)(b*)");
    const found = [...re.matches("aab a aabbb")].map((m) => m.text);
    assert.deepEqual(found, ["aab", "a", "aabbb"]);
    const groups = [...re.captureMatches("aab a")];
    assert.equal(groups.length, 2);
    assert.equal(groups[0].get(1)?.text, "aa");
    assert.equal(groups[1].get(2)?.text, "");
});

test("replacement", () => {
    const re = new Regex("(a+)(b*)");
    assert.equal(re.replaceAll("aab a aabbb", "<\\2\\1>"), "<baa> <a> <bbbaa>");
    assert.equal(re.replaceAll("aab a aabbb", "&", 1), "aab a aabbb");
    assert.equal(re.replaceAll("aab a aabbb", "x", 2), "x x aabbb");
    assert.equal(re.replaceAllWith("aab a", (c: Captures) => c.get(0)!.text.toUpperCase()), "AAB A");
    assert.equal(re.replaceAllWith("zzz", () => "never"), "zzz");
});

test("offsets are byte offsets of the UTF-8 encoding", () => {
    const re = new Regex("b+");
    const m = re.find("éb");
    assert.ok(m);
    assert.equal(m.start, 2);
    assert.equal(m.end, 3);
    assert.equal(m.text, "b");
    const raw = re.find(new Uint8Array([0xff, 0x62, 0x62]));
    assert.ok(raw);
    assert.equal(raw.start, 1);
    assert.deepEqual(raw.bytes, new Uint8Array([0x62, 0x62]));
});

test("options", () => {
    assert.equal(new Regex("ab+", { caseInsensitive: true }).test("xABBx"), true);
    assert.equal(new Regex("ab+").test("xABBx"), false);
    assert.equal(new Regex("^b").test("a\nb"), false);
    assert.equal(new Regex("^b", { newlineSensitive: true }).test("a\nb"), true);
    assert.equal(new Regex("a+", { shortestMatch: true }).find("aaa")?.text, "a");

    const yesNo = new Regex("a", { noCaptures: true });
    assert.equal(yesNo.test("a"), true);
    assert.throws(() => yesNo.find("a"), (e: unknown) => e instanceof RegexError && e.kind === "no-captures");
});

test("compilation errors carry a kind and an offset", () => {
    const kind = (pattern: string, want: string) =>
        assert.throws(() => new Regex(pattern), (e: unknown) => e instanceof RegexError && e.kind === want);
    kind("(a", "paren");
    kind("[a", "bracket");
    kind("a{1", "brace");
    kind("[[:nope:]]", "character-class");
    kind("[[.xx.]]", "collating-element");
    kind("[z-a]", "range");
    kind("a{3,1}", "interval");
    kind("*a", "repeat");
    assert.throws(() => new Regex("a("), (e: unknown) => e instanceof RegexError && e.kind === "pattern" && e.offset === 2);
    try {
        new Regex("a\\");
        assert.fail("expected an error");
    } catch (e) {
        assert.ok(e instanceof RegexError);
        assert.equal(e.kind, "escape");
        assert.match(e.message, /backslash/);
    }
});

test("locales", () => {
    const cs = Locale.open("cs");
    assert.ok(cs);
    assert.equal(Locale.open("no-such-locale"), null);
    assert.ok(Locale.names().includes("cs"));
    assert.ok(embeddedLocaleData().length > 0);

    // In Czech collation, "ch" is one collating element, so [[.ch.]] matches both bytes.
    const re = new Regex("[[.ch.]]", { locale: cs });
    const m = re.find("xchx");
    assert.ok(m);
    assert.equal(m.text, "ch");

    const tr = Locale.open("tr");
    assert.ok(tr);
    assert.equal(new Regex("i", { caseInsensitive: true, locale: tr }).test("I"), false);
    assert.equal(new Regex("i", { caseInsensitive: true, locale: tr }).test("İ"), true);
});

test("contract", () => {
    const re = new Regex("(a|b)*c");
    const c = re.contract(1000);
    assert.equal(typeof c.hasOnePass, "boolean");
    assert.equal(typeof c.hasSolver, "boolean");
    assert.ok(c.heapBytes > 0n);
    assert.ok(c.stackBytes > 0n);
    assert.ok(c.steps > 0n);
    assert.equal(DUP_MAX, 255);

    const onePass = new Regex("(abc+)").contract(1000);
    assert.equal(onePass.hasOnePass, true);
    assert.equal(onePass.hasSolver, false);
    assert.equal(onePass.heapBytes, 37_757n);
    assert.equal(onePass.stackBytes, 6_144n);
    assert.equal(onePass.steps, 937_980n);

    const solver = new Regex("(a|ab)(c|bcd)(d*)").contract(1000);
    assert.equal(solver.hasOnePass, false);
    assert.equal(solver.hasSolver, true);

    const matcherOnly = new Regex("a*").contract(1000);
    assert.equal(matcherOnly.hasOnePass, false);
    assert.equal(matcherOnly.hasSolver, false);
});

test("numeric arguments must be integers", () => {
    const re = new Regex("a");
    assert.throws(() => re.replaceAll("aaa", "x", 1.5), RangeError);
    assert.throws(() => re.replaceAll("aaa", "x", NaN), RangeError);
    assert.throws(() => re.contract(1.5), RangeError);
    assert.equal(new Regex("(a)(b)").captures("ab")?.get(0.5), null);
    const blob = embeddedLocaleData();
    blob[0] ^= 0xff;
    assert.ok(Locale.open("cs"));
});

test("interval bounds", () => {
    assert.equal(new Regex("a{2,3}").find("aaaa")?.text, "aaa");
    assert.throws(() => new Regex(`a{0,${DUP_MAX + 1}}`), RegexError);
});
